File: src/main/java/com/ddc/yieldster/jsonbuilder/service/impl/HelloWorldImpl.java
package com.ddc.yieldster.jsonbuilder.service.impl;

import com.ddc.yieldster.jsonbuilder.model.Advisor;
import com.ddc.yieldster.jsonbuilder.model.ConditionalStep;
import com.ddc.yieldster.jsonbuilder.model.MoveStep;
import com.ddc.yieldster.jsonbuilder.service.HelloWorldService;
import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.google.common.collect.ImmutableList;
import com.swagger.client.codegen.rest.api.SdkServiceApi;
import com.swagger.client.codegen.rest.model.SDKResponse;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.Map;


@Service
public class HelloWorldImpl implements HelloWorldService {
    private static final ObjectMapper OBJECT_MAPPER = new ObjectMapper();


    @Autowired
    private SdkServiceApi sdkServiceApi;

    //Function to get NAV of the provided vault
    //@Param: vaultAddress:- Address of the required vault
    //@Param: date( DD-MM-YYYY HH:MM:SS format) / timestamp. if null uses current blockNumber by default
    //@Param: boolean:- true if provided above param is date; false if the provided param is timestamp
    public SDKResponse getVaultNAV(String vaultAddress) {
        return sdkServiceApi.getVaultNAV(vaultAddress, null, null);
    }

    //Function to get list of invested asset of the provided vault
    //@Param: vaultAddress:- Address of the required vault
    public SDKResponse getVaultAssets(String vaultAddress) {
        return sdkServiceApi.getVaultAssets(vaultAddress);
    }

    //Function to get price of token( stable coin or vault token)
    //@Param: tokenAddress:- Address of the required token
    //@Param: boolean:- true: if the provided tokenAddress is a vault token, false:- if the tokenAddress is stable coin
    //@Param: date( DD-MM-YYYY HH:MM:SS format) / timestamp. if null uses current blockNumber by default
    //@Param: boolean:- true if provided above param is date; false if the provided param is timestamp
    public SDKResponse getTokenPrice(String tokenAddress) {
        return sdkServiceApi.getTokenPrice(tokenAddress, false, null, null);
    }

    /*
    This advisor checks if the increase in price of yvCurve_FRAX over the period of 10 days is above a certain
    threshold (in this case 0.5) the available investable asset (AD) present in vault is swapped to get yvCurve_FRAX
     * */
    public String generateMaximizeAssetReturn() throws JsonProcessingException {

        String yvCurve_FRAX = "0xB4AdA607B9d6b2c9Ee07A275e9616B84AC560139";
        SDKResponse currentTokenData = sdkServiceApi.getTokenPrice(yvCurve_FRAX, false, null, null);
        SDKResponse previousTokenData = sdkServiceApi.getTokenPrice(yvCurve_FRAX, false, "1650737071", false);
        if (currentTokenData.getStatusCode() == 200 && previousTokenData.getStatusCode() == 200) {
            Map<String, Object> currentTokenDataMap = (Map<String, Object>) currentTokenData.getData();
            Map<String, Object> previousTokenDataMap = (Map<String, Object>) previousTokenData.getData();

            Long currentTokenPrice = (Long) currentTokenDataMap.get("Token Price");
            Long previousTokenPrice = (Long) previousTokenDataMap.get("Token Price");
            // The difference in price percentage is calculated.
            double priceDifferencePercentage = ((currentTokenPrice - previousTokenPrice) / previousTokenPrice) * 100;
            if (priceDifferencePercentage > 0.5) {
                Advisor advisor = Advisor.builder()
                        .type("Sample_Advisor")
                        .steps(ImmutableList.of(
                                MoveStep.builder()
                                        .fromAsset("AD")
                                        .toAsset(yvCurve_FRAX)
                                        .build())).build();
                return OBJECT_MAPPER.writerWithDefaultPrettyPrinter().writeValueAsString(advisor);
            }
        }
        return null;
    }

    //Sample Advisor: Maximize APR
    //Here the variable used like AD, TMin ... are predefined ones that are recognizable by the Yieldster Application and the corresponding value is
    // replaced during it's execution
    //In this type of Advisor most of the computation is done in the backend
    private Advisor createAdvisorMaximizeAPR() {
        return Advisor.builder()
                .type("Maximize_APR")
                .steps(ImmutableList.of(
                        ConditionalStep.builder()
                                .conditionalStatement("AD >= TMin")
                                .trueExecuteStep(ImmutableList.of(
                                        ConditionalStep.builder()
                                                .conditionalStatement("APRMax <= APRInv")
                                                .trueExecuteStep(ImmutableList.of(
                                                        MoveStep.builder()
                                                                .fromAsset("AD")
                                                                .toAsset("APInv")
                                                                .build()
                                                ))
                                                .falseExecuteStep(ImmutableList.of(
                                                        ConditionalStep.builder()
                                                                .conditionalStatement("IDMax * BTotal > Pin + Pout - Gas")
                                                                .trueExecuteStep(ImmutableList.of(
                                                                        MoveStep.builder()
                                                                                .fromAsset("(APInv + AD)")
                                                                                .toAsset("APMax")
                                                                                .build()
                                                                ))
                                                                .falseExecuteStep(ImmutableList.of(
                                                                        MoveStep.builder()
                                                                                .fromAsset("AD")
                                                                                .toAsset("APInv")
                                                                                .build()
                                                                ))
                                                                .build()
                                                ))
                                                .build()
                                ))
                                .falseExecuteStep(ImmutableList.of(
                                        ConditionalStep.builder()
                                                .conditionalStatement("APRMax > APRInv")
                                                .trueExecuteStep(ImmutableList.of(
                                                        ConditionalStep.builder()
                                                                .conditionalStatement("IDMax * BTotal > Pin + Pout - Gas")
                                                                .trueExecuteStep(ImmutableList.of(
                                                                        MoveStep.builder()
                                                                                .fromAsset("(APInv + AD)")
                                                                                .toAsset("APMax")
                                                                                .build()
                                                                ))
                                                                .falseExecuteStep(ImmutableList.of(
                                                                        MoveStep.builder()
                                                                                .fromAsset("AD")
                                                                                .toAsset("APinv")
                                                                                .build()
                                                                ))
                                                                .build()
                                                ))
                                                .build()
                                ))
                                .build(),
                        ConditionalStep.builder()
                                .conditionalStatement("active AP (APINV) is disabled")
                                .trueExecuteStep(ImmutableList.of(
                                        MoveStep.builder()
                                                .fromAsset("(APInv + AD)")
                                                .toAsset("APMax")
                                                .build()
                                ))
                                .falseExecuteStep(null)
                                .build()
                ))
                .build();
    }
}

File: Yieldster-SDK/Service/Impl/HelloWorldImpl.py
import json

import null

from Models.Advisor import Advisor
from Models.ConditionalStep import ConditionalStep
from Models.MoveStep import MoveStep

from SdkServiceApi import sdkApi


# @zope.interface.implementer(Service.HelloWorldService)
class HelloWorldImpl:

    def generate_maximize_asset_return(self):
        sdk = sdkApi()
        currentTokenData = sdk.get_token_price(token_address="0xdAC17F958D2ee523a2206206994597C13D831ec7",
                                               is_vault_token=False)
        previousTokenData = sdk.get_token_price(token_address="0xdAC17F958D2ee523a2206206994597C13D831ec7",
                                                is_vault_token=False,
                                                time_stamp="1653974358", is_date=False)

        if currentTokenData['status_code'] == 200 and previousTokenData['status_code'] == 200:
            currentTokenDataMap = currentTokenData['data']
            previousTokenDataMap = previousTokenData['data']
            currentTokenPrice = currentTokenDataMap['Token Price']
            previousTokenPrice = previousTokenDataMap['Token Price']
            priceDifferencePercentage = ((currentTokenPrice - previousTokenPrice) / previousTokenPrice) * 100
            if priceDifferencePercentage > 0.5:
                advisor = Advisor()
                advisor.advisorType = "Sample_Advisor"
                movestep = MoveStep()
                movestep.toAsset = "0xdAC17F958D2ee523a2206206994597C13D831ec7"
                movestep.fromAsset = "AD"
                advisor.steps.append(movestep.__dict__)
                return json.dumps(advisor.__dict__)
        return None

    def createAdvisorMaximizeAPR(self):
        advisor = Advisor()
        advisor.advisorType = "Maximize_APR"
        conditional_step1 = ConditionalStep()
        conditional_step1.conditionalStatement = "AD >= TMin"
        conditional_step2 = ConditionalStep()
        move_step1 = MoveStep()
        conditional_step2.conditionalStatement = "APRMax <= APRInv"
        move_step1.fromAsset = "AD"
        move_step1.toAsset = "APInv"
        conditional_step2.trueExecuteStep.append(move_step1.__dict__)
        conditional_step3 = ConditionalStep()
        conditional_step3.conditionalStatement = "IDMax * BTotal > Pin + Pout - Gas"
        move_step1.fromAsset = "(APInv + AD)"
        move_step1.toAsset = "APMax"
        conditional_step3.trueExecuteStep.append(move_step1.__dict__)
        move_step1.fromAsset = "AD"
        move_step1.toAsset = "APInv"
        conditional_step3.falseExecuteStep.append(move_step1.__dict__)
        conditional_step2.falseExecuteStep.append(conditional_step3.__dict__)
        conditional_step1.trueExecuteStep.append(conditional_step2.__dict__)
        conditional_step1.falseExecuteStep.append(conditional_step3.__dict__)
        advisor.steps.append(conditional_step1.__dict__)
        advisor1 = json.dumps(advisor.__dict__)
        return advisor1
